### ISSUE TITLE
Compile to WASM

### ISSUE THREAD

**Issue by GarboMuffin** [TurboWarp/scratch-vm#16]
It may be possible to compile a subset of Scratch blocks to webassembly for performance.

**Comment by mguinhos** [comment]
Consider AssemblyScript as a target. It is TypeScript compiled to WebAssembly https://www.assemblyscript.org/

**Comment by mollthecoder** [comment]
Another possible thing to consiter is c, c++, or rust because they are popular with webassembly and receive the most attention when it comes to develoupment, leading to webassembly enhancement and better documention.

**Comment by GarboMuffin** [comment]
In the context of this project we would have to generate WASM bytecode dynamically at runtime. Adding a dependency on rustc at runtime is not viable for us

**Comment by mollthecoder** [comment]
That might take a long time to calculate especally for large projects that turbowarp is so famous for.

**Comment by mollthecoder** [comment]
I had a great idea

**Comment by mollthecoder** [comment]
What if you prepared pre-compiled webassembly functions for each block?

**Comment by mollthecoder** [comment]
And then simply embeded using webassembly API

**Comment by GarboMuffin** [comment]
Compiling to WASM is entirely theoretical at this point and it's not clear whether it would actually be any faster because WASM <-> JavaScript interop is not free, and [WebAssembly is not a magic cure to all performance issues](https://surma.dev/things/js-to-asc/index.html). Browser's *really* don't want websites to syncronously compile WASM, which is a real problem for an app like TurboWarp that dynamically generates code at runtime and has to execute it immediately or else projects misbehave.

The JavaScript that TurboWarp generates is still significantly slower than the same code written by a human. We have a ways to go before WASM will be necessary to go faster.

**Comment by mollthecoder** [comment]
That is what hand-optimization is for according to the article you linked. Satifying the user takes hours and hours of hard, tough work. Because users don't know that in real life coding is a lot harder than snapping together blocks.

**Comment by mguinhos** [comment]
The most easy solution is creating turbowarp's own Bytecode, and then running it inside a VM/Engine implemented in WebAssembly (C/C++)

it will be faster than compiling webassembly modules dinamically or converting to JS
Also execution should be faster than the current "compiled" JS code

For those who don't know, WebAssembly modules are not interpreted, they are compiled AOT to machine code before execution.

So a TurboWarp VM/Engine should be a pratical ideia

I am currently working on a prototype...

@GarboMuffin

**Comment by GarboMuffin** [comment]
A bytecode interpreter in WASM will not necessarily be faster than letting the browser's JavaScript JIT figure out how to optimize our scripts

**Comment by mguinhos** [comment]
Because javascript runtimes arent really standard and their behaviour change depending on the runtimr and the execution context.

There is so much entropy

JS is a dynamic and object oriented language. in a custom bytecode we can have guarantees that we cannot have in JS code.

WASM is very efficient, near native and sandboxed.

As a personal opinion, i think the actual Scratch->JS compiler is still very inefficient

**Comment by GarboMuffin** [comment]
Everything you said is correct but it doesn't prove that a bytecode interpreter in WASM will outperform our JavaScript as interpreting bytecode has a very non-insignificant performance overhead.

I am curious how well it performs. Do let me know when you having a working prototype to benchmark.

**Comment by mguinhos** [comment]
> A bytecode interpreter in WASM will not necessarily be faster than letting the browser's JavaScript JIT figure out how to optimize our scripts
> 
It will be faster if well designed

Compiling WASM modules at runtime is not pratical because of the specifications of the binary format. The encoding of WASM modules are heavy and inneficient for code generation.

The only alternative is creating your own VM inside WASM

**Comment by mguinhos** [comment]
> Everything you said is correct but it doesn't prove that a bytecode interpreter in WASM will outperform our JavaScript as interpreting bytecode has a very non-insignificant performance overhead.
> 
> I am curious how well it performs. Do let me know when you having a working prototype to benchmark.

Sure!

**Comment by IvanikuV2** [comment]
Any updates on this?

**Comment by Tacodiva** [comment]
I have been working on a Scratch -> WebAssembly compiler ([here](https://github.com/Tacodiva/Catnip/)) for almost a year now (on and off), and have worked on the TurboWarp compiler, and I think I am finally qualified to answer the question of if TurboWarp can improve performance with WASM.

I hope this doesn't come off as pessimistic, but to me, it seems like there is very likely no decent way to implement WebAssembly into TurboWarp and get a performance benefit from it. The problem is, as you've stated, the WASM <-> JavaScript interop kills performance.

In my compiler, I have been seeing performance gains over TurboWarp only because I have built the whole thing around WASM. I have been implementing the runtime in C and compiling it to WASM to avoid JS as much as possible. Looking at the TurboWarp ecosystem, this is really not a possibility. scratch-vm needs to have so much information on the JS side to pass to the GUI, the renderer, extensions, etc. that without rewriting basically everything, I think WASM will be slower 99% of the time. The fact is, V8 is really good at its job. I had to do quite a bit of work to beat it even in my almost entirely WASM runtime. The idea that extension compatibility and editor compatibility could be maintained while compiling to WASM seems to me to be almost impossible.

I can also say with almost 100% certainty that a scratch bytecode interpreter compiled to wasm (as suggested in this issue) would yield very poor performance. There's no way that an interpreter like that would be able to beat V8.

**Comment by GarboMuffin** [comment]
Closing per above

**Comment by LuYifei2011** [comment]
Quick update: I’ve been experimenting with a WASM execution kernel for scratch-vm, and the current prototype now outperforms TurboWarp on several workloads in my current benchmark suite, although a few workloads are still slower and are being optimized.

This is not just a WASM bytecode interpreter. It currently uses a hybrid compiled/interpreted execution model inside WASM, keeps much of the runtime state resident in WASM memory, and batches JS/renderer interactions to reduce interop overhead.

| Item | WASM  | TurboWarp | Relative |
|------|-------:|------------:|------:|
|PRNG|50703  | 22511|2.25x|
|Terrain Generator|62969|60500|1.04x|
|Raycaster|48572|32668|1.49x|
|Quicksort|14273|18680|0.76x|
|List Search|6736|1233|5.46x|
|String Parsing|20945|14115|1.48x|
|Triangle Filler|8049|7252|1.11x|
|Pen Text Engine|8115|9642|0.84x|
|Character set stress test.sb3 (lower is better)|0.24s|0.33s|1.38x faster|

[Scratch Benchmark (v1.0.2)_fixed_pen 100x.html](https://github.com/user-attachments/files/31092163/Scratch.Benchmark.v1.0.2._fixed_pen.100x.html)
[Scratch Benchmark (v1.0.2)_fixed_pen 100x.sb3.zip](https://github.com/user-attachments/files/31092164/Scratch.Benchmark.v1.0.2._fixed_pen.100x.sb3.zip) (based on https://scratch.mit.edu/projects/441947766)
[Character set stress test.html](https://github.com/user-attachments/files/31092513/Character.set.stress.test.html)
[Character set stress test.sb3.zip](https://github.com/user-attachments/files/31092515/Character.set.stress.test.sb3.zip)

(GitHub does not allow direct .sb3 uploads, so the files are wrapped as .zip; rename them back to .sb3 after downloading.)

The prototype originally grew out of a separate project-protection experiment, so the current build also contains unrelated signing/obfuscation machinery, and the interpreted fallback was partly introduced for that use case. Those parts are not fundamental to the WASM performance approach and could be removed or simplified.

The implementation is still experimental and has not been published yet. I’m still working on several optimizations and plan to post a more detailed architecture/benchmark report later. If a maintainer is interested before then, I can provide the current uncleaned source  for review.

@GarboMuffin

**Comment by GarboMuffin** [comment]
Happy for you

**Comment by GarboMuffin** [comment]
The open source licenses used by various parts of TurboWarp may obligate you to provide certain portions of the source code to anyone receiving a compiled version of it. You can read the various LICENSE files to learn more about that.

**Comment by LuYifei2011** [comment]
Thanks for the reminder. The archive currently contains the modified MPL-covered TurboWarp components used by these builds. The separate compiler/protection tooling is not included because it is maintained independently from those source trees.

[scratch-render.zip](https://github.com/user-attachments/files/31094990/scratch-render.zip)
[tw-packager.zip](https://github.com/user-attachments/files/31094991/tw-packager.zip)
[scratch-vm.zip](https://github.com/user-attachments/files/31095017/scratch-vm.zip)

**Comment by LuYifei2011** [comment]
One thing I found while profiling the WASM prototype is that it currently performs more aggressive static analysis and carries stronger type information through compilation than TurboWarp’s JS compiler in some cases.

For example, when the WASM compiler can prove that both operands of a comparison are numeric, it can emit a direct numeric comparison. TurboWarp currently still goes through generic helpers such as `compareLessThan` / `compareGreaterThan` in cases like this, preserving the dynamic coercion and slow paths required by general Scratch semantics.

Conceptually:

```js
// generic
compareLessThan(left, right)
// when both operands are statically proven numeric
left < right
```

In one comparison-heavy workload, these generic comparison helpers and their slow paths account for a significant portion of TurboWarp’s execution time. This seems to be one reason the WASM implementation performs particularly well on some algorithm shapes.

I think some of these static-analysis ideas could potentially be useful to TurboWarp’s existing JS compiler independently of the WASM execution kernel. I’m planning to test the comparison specialization separately first, while preserving Scratch’s behavior for numeric strings, whitespace, NaN, hexadecimal strings, etc.

**Comment by GarboMuffin** [comment]
We very much do try to do that. What cases do you think are missing?

**Comment by LuYifei2011** [comment]
I should clarify that the workload I was looking at was a quicksort project, so the case I had in mind may actually be earlier type propagation rather than the comparison codegen itself.

In particular, `item of list` may still be typed as ANY in cases where whole-project analysis could potentially prove the list is numeric. I’m checking current develop now to confirm the exact missing case before making a stronger claim.

**Comment by LuYifei2011** [comment]
I checked this more carefully. The difference I was seeing is indeed mostly earlier type analysis/propagation, rather than comparison code generation itself.

TurboWarp already specializes comparisons when the IR can prove the operands are numeric. The WASM compiler currently goes further by doing more whole-project analysis before reaching that point.

One concrete example is list element types. In TurboWarp, item of list is generally typed as ANY, so a comparison using two list items still needs a dynamic path. The WASM compiler analyzes the reachable writes to each list and, when it can prove that a list only contains numbers, treats it effectively as number[]. That lets item of list remain numeric through lowering and subsequent comparisons become direct f64 comparisons.

It does similar analysis across procedure calls and also distinguishes some string cases more precisely; for example, if a value can be proven to be a non-numeric string, the WASM side can skip Scratch’s numeric-conversion path entirely and go directly to string comparison.

So I think the potentially transferable optimization is not another comparison fast path, but stronger propagation of type information across things like list accesses and procedure boundaries. A relatively conservative first version for the JS compiler could keep normal JS arrays, but infer element types from all statically reachable writes and let item of list return a more specific IR type when that proof is safe.

This also explains why the WASM advantage is workload-dependent: on purely numeric data, V8’s native JS number/array path can actually be faster, while the stronger analysis helps much more when it avoids repeated generic Scratch coercion for strings or mixed-looking values.

**Comment by Tacodiva** [comment]
I've thought about that optimisation for lists before and it could *maybe* be done in TurboWarp on the project page, but in the editor recompiling the whole project when any script is changed or if a string item is added to a list manually would be difficult. Another problem is extensions could modify lists and add strings which idk how to deal with. 
 
Basically it would be quite difficult and require quite a restructure to do this in TW.

**Comment by GarboMuffin** [comment]
even outside of the editor and without extensions,

<img width="499" height="411" alt="Image" src="https://github.com/user-attachments/assets/d002cb4d-6d8d-4fab-9b0e-d31672f63f39" />

<img width="242" height="127" alt="Image" src="https://github.com/user-attachments/assets/75a3da80-e90b-4bb5-b82d-266236a4bd28" />

**Comment by GarboMuffin** [comment]
you could reliably track list content types after a "delete all" block runs. that probably helps a lot in some benchmarks that are reliably setting up clean environments between runs, especially if you can track that down through function calls too, but not clear how much it helps in real projects

**Comment by GarboMuffin** [comment]
being able to import/export lists in our gui without being in the editor is probably a misfeature at this point too though

**Comment by Tacodiva** [comment]
> even outside of the editor and without extensions,

Not that this optimisation is viable, but I was imagining if the list was shown or there were any show/hide list blocks the optimisation would be disabled

**Comment by CST1229** [comment]
> I've thought about that optimisation for lists before and it could _maybe_ be done in TurboWarp on the project page, but in the editor recompiling the whole project when any script is changed or if a string item is added to a list manually would be difficult. Another problem is extensions could modify lists and add strings which idk how to deal with.

MAYBE if some way is figured out to allow compiled scripts to drop back down to the interpreter without restarting them?

**Comment by LuYifei2011** [comment]
> The implementation is still experimental and has not been published yet. I’m still working on several optimizations and plan to post a more detailed architecture/benchmark report later. If a maintainer is interested before then, I can provide the current uncleaned source for review.

I used Codex to do a static review of the current implementation and organize the architecture, execution flow, state ownership, performance-related design, Git history, technical debt, and possible upstream-relevant ideas into a more complete report.

Some parts are specific to the WASM/protected-project architecture, but there are also several optimizations that do not fundamentally depend on WASM and might be useful to TurboWarp independently — for example precise input dependency tracking/caching, Pen batching and renderer instancing, collision geometry preprocessing, and some target/thread data-structure optimizations.

Section 10 is probably the most relevant part from an upstream perspective, as it separates WASM-specific mechanisms from ideas that may also apply to TurboWarp's existing JS/runtime/renderer paths.

I'm posting it mainly as a reference in case any of these ideas are useful upstream. The report distinguishes source-backed facts, inferences, and suggestions, and also notes the limitations of the static review.

<details>
<summary>Full architecture / implementation report</summary>

# Current WASM Approach: Architecture, Execution Flow, and Design Evolution

> Baseline: static review of the 2026-08-17 workspace. The main repository HEADs were scratch-vm `2097b28a`, obfuscator `8c8c067`, scratch-render `67bb559f`, and tw-packager `35d7946`. No runtime code or project structure was modified; only this report was added.
>
> Evidence labels: **[Fact]** is directly supported by the current source code, tests, or Git history; **[Inference]** is an explanation derived from multiple pieces of evidence; **[Suggestion]** is an entry point for future discussion and is not presented as current behavior. The full browser benchmark or end-to-end test suite was not run, so “a test exists” does not mean “it was verified locally in this review.”

## 0. Executive Summary

**[Fact]** The current approach does not “replace the Scratch VM with WASM.” The existing TurboWarp/Scratch VM remains the host for loading, scheduling, extensions, rendering, and compatibility, while portable optimized IR is lowered into a private `program.wasm`. WASM is responsible for compiled scripts, variables/lists, thread-private state, PRNG, part of the target authority, string/numeric operations, and some collision queries. JavaScript still owns Scratch thread objects, event scheduling, extensions, visible VM/renderer objects, and host semantics that cannot or should not be moved into WASM.

**[Fact]** The release artifact is a signed `.sb3c2`: `project.json` (the public project with source removed) + `program.wasm` (private program and state) + `effects.wasm` (fixed-memory, import-free effects transport) + optional `online.json` + integrity manifest. During loading, the archive is verified first, then both WASM modules are compiled and instantiated, and finally each hat’s compilation cache is replaced with a `WasmThreadAdapter`.

**[Fact]** The ABI is now **6** and requires the Wasm features `bulk-memory`, SIMD, tail calls, and typed function references. However, source files still retain names such as `abi5-*`, the class name `ABI5EffectsChannel`, and the `benchmark:abi5` command. This is naming debt left over from ABI evolution, not two ABIs running in parallel.

**[Inference]** The greatest complexity in the system is currently not lowering itself, but precise synchronization between “private WASM authority” and the existing visible JS target/renderer authority. `compiled-project.js`, `wasm-thread-adapter.js`, `abi5-effects-channel.js`, and `rendered-target.js` together form a cross-layer state machine. This preserves semantics, but also means that adding an opcode, clone behavior, Pen behavior, fencing, collision logic, or extension behavior can require coordinated changes in several places.

**[Fact]** Recent optimizations are concentrated around three boundary costs: reducing JS↔WASM crossings (precise input masks, batched snapshots, compact exports); batching large amounts of motion/Pen work into ordered effects; and preprocessing renderer alpha silhouettes into conservative geometry so that WASM can perform tri-state fast collision queries, with unknown cases falling back to the renderer’s exact path.

## 1. Scope, Repositories, and Architecture Map

### 1.1 Workspace Layers

| Layer / repository | Current responsibility | Boundary with the WASM approach |
| --- | --- | --- |
| `ScratchCN_V4_scratch-vm/` | Modified Scratch VM: SB3 loading, IR generation, thread/scheduler integration, `sb3c2` installation, JS/WASM adapter, target/extension glue | The only runtime host; does not generate the final WASM bytecode |
| `ScratchCN_v4_code-obfuscation/` | Protector: validates build input, performs IR lowering, generates/verifies `program.wasm` and `effects.wasm` | Does not own the browser VM, renderer, or live Scratch targets; consumes portable data only |
| `scratch-render/` | WebGL renderer, Pen batching, sprite instancing, collision geometry descriptors, fencing | Knows nothing about IR/WASM; only exposes renderer extensions used by the VM bridge |
| `tw-packager/` | Reads SB3, invokes VM+protector, signs/packages `.sb3c2`, generates player, benchmark/diagnostics | Compilation orchestration and publishing boundary; does not interpret IR or execute WASM |
| `benchmark-results/`, `tw-packager/docs/benchmark-results/` | Archived performance records | Evidence/experimental material; not part of runtime execution |
| `TurboWarp-issue-18/source code/` | A mirrored snapshot / issue-related material | **Not treated as the current main implementation** unless a later manual review confirms its consumer |

### 1.2 Key Directories and Files

| Module | Key files | Responsibility and boundary |
| --- | --- | --- |
| VM IR | `ScratchCN_V4_scratch-vm/src/compiler/compile.js`, `irgen.js`, `iroptimizer.js`, `intermediate.js`, `enums.js` | Existing TurboWarp JS AOT IR frontend; `compile.toIR()` is the WASM build entry point. The IR model is shared with JS AOT; WASM is not generated here. |
| Portable IR | `compiled-ir-serializer.js`, `opcode-semantics.js` | Converts `Intermediate*` objects into pure data, rejecting functions, cyclic references, and prototype objects; defines the VM semantic inventory of 77 stack + 90 input = 167 opcodes. |
| VM build/load | `compiled-project.js`, `compiled-project-integrity.js`, `virtual-machine.js` | Exports build input, validates protected artifacts, loads zip archives, verifies integrity before execution, instantiates WASM, and binds targets/hats/renderer lifecycle. |
| VM↔WASM adaptation | `wasm-thread-adapter.js`, `abi5-effects-channel.js`, `wasm-kernel-abi5.js`, `wasm-effects-abi5.js` | `WasmThreadAdapter` converts `runSlice` statuses into Thread/Sequencer behavior; effects channel decodes, batches, and falls back; the latter two files hold the ABI 6 contract/probes and effects shape validation. |
| Target synchronization | `sprites/rendered-target.js`, `extensions/scratch3_pen/index.js`, `engine/runtime.js` | JS target mutation revision, collision dirty tracking, host projection of batched effects, tick-end drain. These files contain direct knowledge of the compiled runtime. |
| Protector orchestration | `ScratchCN_v4_code-obfuscation/src/pipeline.mjs`, `index.mjs`, `abi6-features.mjs` | Fail-closed validation of build-input/feature/opcode contracts, randomized layout seed, lowering invocation, artifact/diagnostics/sidecar assembly. |
| Protector lowering | `ir-lowering.mjs` | ~6k-line core lowering implementation: private memory layout, value representation, opcode lowering, procedure/suspension state machine, host commands, clone/list/string specializations, collision planning. |
| Wasm emitter | `wasm-kernel.mjs`, `effects-kernel.mjs`, `effects-kernel.wat` | Hand-written binary encoding; creates `program.wasm` ABI exports/imports and the fixed-memory effects module; validates generated output. |
| Renderer | `scratch-render/src/RenderWebGL.js`, `PenSkin.js`, `Drawable.js`, `CollisionGeometry.js` | `penLineRuns`, batched fencing, contiguous sprite instancing, renderer-authoritative collision descriptor/state. |
| Packager | `tw-packager/src/packager/compile-project.js`, `packager.js`, `download-project.js`, `load-project.js`, `abi6-features.js` | Dynamically loads dependencies, freezes remote extensions, invokes the protector, signs/packages output, and points the generated player at `loadCompiledProjectPackage`. |
| Benchmark/diagnostics | `tw-packager/src/benchmark/abi5-runner.js`, `scripts/benchmark-abi5.js`, `docs/abi5-benchmark.md` | Compares JS AOT against ABI6 WASM; GPU fencing, profiling, collision smoke tests, instrumentation. Names still retain ABI5. |

### 1.3 High-Level Relationship Diagram

```mermaid
flowchart LR
  SB3[".sb3 / Scratch project"] --> P["tw-packager\ncompileProject"]
  P --> VM1["scratch-vm\nloadProject + compile.toIR"]
  VM1 --> BI["compiled IR v2 build input\npublicProject + privateState + contracts"]
  BI --> O["obfuscator/protector\nlowerProgram + Wasm emitter"]
  O --> ART[".sb3c2\nproject.json + program.wasm + effects.wasm + manifest"]
  ART --> PLAYER["packager player / scaffolding"]
  PLAYER --> VM2["VM loadCompiledProjectPackage\nvalidate + instantiate + install"]
  VM2 --> ADAPTER["WasmThreadAdapter\nrunSlice / host bridge"]
  ADAPTER <--> WASM["program.wasm\nprivate state + compiled scripts"]
  WASM --> FX["effects.wasm\nfixed effect memory"]
  ADAPTER --> FXC["ABI5EffectsChannel\ndrain / batch / fallback"]
  FXC --> R["scratch-render\nPenSkin / RenderWebGL"]
  VM2 <--> R
```

The ideal main dependency direction is `packager → VM/obfuscator → renderer`; the actual runtime has reverse edges: the VM’s compiled bridge actively reads renderer state to synchronize collision, while renderer capabilities determine whether effects fast paths are available.

## 2. Complete Execution Flow: From SB3 to Screen

### 2.1 Build and Packaging

1. **Read the project.** `tw-packager/src/packager/compile-project.js::compileProject()` creates `VM` and `Storage`, forces custom extensions into the worker sandbox, enables the compiler, and then calls `vm.loadProject(projectData)`. **[Fact]**
2. **Share the IR frontend.** `VM.exportCompiledProjectBuildInput()` calls `compiledProject.create()`; that function walks hats on `runtime.executableTargets`, creates temporary `Thread` objects, and obtains IR through `compile.toIR(thread)` (`IRGenerator → IROptimizer`). Normal JS AOT still uses the same `compile.js` with `JSGenerator`. **[Fact]**
3. **Split public/private state.** `compiledProject.create()` places variables, lists, and their initial values into `privateState`; project JSON, costume/monitor/extension information, and the `compiledRuntime` feature contract go into `publicProject`. IR uses deterministic ordinals from executable targets so the private layout cannot drift with runtime target ordering. **[Fact]**
4. **Portable build input.** `compiled-ir-serializer.js::serializeValue()` encodes `IntermediateStack`, `IntermediateStackBlock`, and `IntermediateInput` as `$kind` trees while rejecting function/symbol/bigint/cycle values; it also attaches opcode inventory and runtime kernel/feature contracts. **[Fact]**
5. **Protection and lowering.** The packager validates the ABI contract and opcode build input, freezes remote extensions into data URLs, then calls `protectCompiledProject()`. The protector validates input and the runtime kernel; `lowerProgram()` builds the private memory image, target/list layouts, WASM script bodies, input masks, collision plans, and host-call descriptors; `createProgramWasm()` and `createEffectsWasm()` generate binaries. **[Fact]**
6. **Release gate.** `compileProject()` requires a signing hook, compares opcode reports from the VM/packager/protector, and rejects production builds when `coverageComplete !== true` (with very narrow exceptions for benchmark/readable-WASM hooks). The artifact sidecar is deleted before `saveCompiledProject()` and cannot enter the signed package. **[Fact]**
7. **Archive.** `VirtualMachine._saveCompiledProjectZip()` writes `project.json`, `program.wasm`, `effects.wasm`, and optional `online.json`; `compiled-project-integrity` creates the manifest. **[Fact]**

### 2.2 Loading, Installation, and Scheduling

1. **Player loading.** The generated packager player enters `VM.loadCompiledProjectPackage()` through `scaffolding.loadCompiledProjectPackage()`; ordinary `.sb3` files still use `loadProject()`. `download-project.js` also checks `project.json`, program/effects modules, the manifest, and feature contracts during package import. **[Fact]**
2. **Validate before execution.** The VM first performs archive path/integrity validation, reads the JSON and the two WASM files; `compiledProject.decodeProgram()` first calls `decodeEffects()` (fixed, import-free effects module), then compiles/links the program and verifies exact import/export shape, ABI, and feature mask. **[Fact]**
3. **Install.** `compiledProject.install()` calls `api.bindTarget(ordinal, 0)` for original targets, calls `resolveScript()` for each hat, and writes `startingFunction` into the Blocks cache. That factory creates a `WasmThreadAdapter` when a Thread starts, so the existing Thread/Sequencer lifecycle remains in use. **[Fact]**
4. **Each scheduling slice.** When necessary, `WasmThreadAdapter.next()` drains pending effects, writes host query results, waits for broadcast children/promises, reconciles externally modified targets, refreshes collision dirty state, stages data according to the input mask, and then calls `api.runSlice(threadHandle, 32768)`. **[Fact]**
5. **Status return.** `runSlice` statuses `DONE/BUDGET/YIELD/YIELD_TICK/HOST_COMMANDS/HOST_PENDING/WAIT_TIME/EFFECTS_READY/ERROR` are mapped into JS Thread status changes, host command execution, monitor synchronization, or effect drains. `BUDGET` does not normally yield Scratch scheduling inside warp; only when the sequencer timer exceeds 500 ms does it become visible. **[Fact]**
6. **Tick finalization/rendering.** After `Sequencer.stepThreads()`, `Runtime._step()` calls the binding’s `flushEffects('tick-end')`, then proceeds to renderer draw. **[Fact]**

### 2.3 Specialized Effects / Pen / Collision Paths

- The **only imports** of `program.wasm` are the five effect functions under `sccn:abi5-effects`; it does not import host state directly from JS. `effects.wasm` itself has no imports and uses fixed 16-page memory as transport between the two WASM modules. **[Fact]**
- `ABI5EffectsChannel.drain()` creates reusable typed-array views over effects memory and validates topology. It can take the safe-motion, clone-graphic-motion, ordered-run, or traditional line-run path. On success it calls `renderer.penLineRuns()`; on failure it falls back item by item to the original `target.setXY()/Pen` semantics and reconciles the actual host result back into WASM. **[Fact]**
- `PenSkin.drawLineRuns()` uses instanced rendering. On WebGL1 without `ANGLE_instanced_arrays`, it retains a per-triangle fallback. `RenderWebGL.fenceLineCoordinatesOfDrawable()` reproduces sequential fencing in batches. **[Fact]**
- For touching queries, the renderer provides immutable conservative alpha geometry and transforms (position removed from the matrix). The VM is responsible for cache/registration/dirty invalidation. WASM returns true/false for trusted geometry and tri-state unknown for unsafe or unsupported paths, with unknown falling back to the renderer. **[Fact]**

## 3. Data and State Ownership

### 3.1 State Ownership Table

| State | Primary authority | Mirrors / consumers | Synchronization boundary and rules |
| --- | --- | --- | --- |
| Compiled variables, lists, private strings, PRNG, loop counters | `program.wasm` private memory | JS only bridges visible monitors/cloud; debug trace can read summaries | Initial values are written from `privateState` into memory; monitor/cloud state is synchronized at adapter boundaries; normal Scratch code should not expose these values back into public JSON. |
| Target identity and clone private state | WASM handle + target layout | JS `WeakMap(target→handle)`, `Map(handle→target)`, ordinal | Originals bind by ordinal; clones deep-copy from the source handle; effects are drained before removal; non-original targets call `bindTarget(-1, handle)` for retirement. |
| `Thread`, hat starts, broadcast-child sets, promises, wait time | JS VM/Sequencer/Thread | WASM stores thread handle, PC/call/argument/scratch/query slots | The adapter maps `runSlice` status back to JS; `broadcast and wait` child Thread sets remain in JS. |
| Position/direction/size/costume/visibility/effects/Pen public object | Phased: WASM can temporarily be authoritative on hot traces; JS target/renderer is the externally visible projection | `RenderedTarget`, Drawable, effects channel, target revisions | Before adapter turns, externally written JS state is reconciled by revision; after WASM effect drains, `_projectCompiledPosition` projects state; fencing/drag/custom-callback fallback results must be written back into WASM. |
| Renderer drawable, GL texture, Pen framebuffer, draw order | `scratch-render` | VM holds drawableID; WASM does not directly own GL state | VM interacts only through renderer APIs; effects channel converts WASM data into renderer submissions. |
| Collision alpha geometry/transform | Descriptor/state generated by renderer | VM cache + WASM geometry handle | Skin/quality changes mark dirty; VM registers/releases geometry and writes transforms/flags; unknown, shape-effect, or unsafe cases fall back to renderer. |
| Input (mouse, keys, timer, username, answer, clock, stage dimensions) | JS IO/runtime | WASM staged slots | `scriptInputMask` stages only required fields; global input uses epoch/cache; clock can be refreshed separately inside private warp budgets. |
| Extension/compat/addon | JS extension system | Lowered host-call descriptor | Executed at `HOST_COMMANDS/HOST_PENDING`; may be asynchronous; result is written into query input before resuming the original PC. |
| Artifact integrity, feature contract, online descriptor | Packager/loader policy | VM validator | Validated before any WASM compile/instantiate operation; `required` online mode is explicitly rejected by the current pipeline. |

### 3.2 Key Synchronization Invariants

**[Fact]** `target._compiledStateRevision` is incremented by JS mutations including `setXY`, `setDirection`, `setVisible`, `setSize`, graphic effects, costume changes, and drag operations. The adapter compares revisions before selectively calling `reconcileTargetIfModified()`. This avoids copying the entire target on every `next()`.

**[Fact]** Position-only commands prefer `reconcileTargetPositionFast()`. Observation-query fast paths can be used when a query value is available; otherwise the full `reconcileTarget()` path is used. `globalInputEpoch/globalInputCacheMask` and per-script input masks reduce repeated staging.

**[Fact]** Effects are not materialized after every WASM instruction. Buffer-ready, target/collision fences, yield/done/wait, target removal, and tick end are explicit drain boundaries. Clone graphic motion also has a deferred/coalesced path.

**[Inference]** This design tightly couples “semantic boundaries” and “performance boundaries”: adding an opcode that observes targets, executes an extension, or changes scheduler-visible state does not only require code generation—it also requires reconsidering command, reconciliation, drain, and yield boundaries.

## 4. Module Dependencies, Cross-Layer Dependencies, and Coupling Hotspots

### 4.1 Main Dependency Directions

```mermaid
flowchart TD
  A["compiler IR/enums\n(scratch-vm)"] --> B["compiled-project build input\n(scratch-vm)"]
  B --> C["opcode contract\n(packager + protector)"]
  B --> D["lowerProgram\n(obfuscator)"]
  D --> E["program/effects WASM"]
  E --> F["compiled-project.decode/install\n(scratch-vm)"]
  F --> G["WasmThreadAdapter"]
  G --> H["ABI5EffectsChannel"]
  H --> I["RenderWebGL / PenSkin"]
  I --> F
  J["tw-packager"] --> B
  J --> D
  J --> F
```

| Relationship | Assessment | Evidence |
| --- | --- | --- |
| VM IR → protector | Reasonable one-way build dependency | Portable `build input` is the cross-repository data contract; the protector does not import scratch-vm. |
| packager → VM + protector | Orchestration-layer dependency | `compile-project.js` dynamically imports `scratch-vm` and the protector and owns signing/extension-freeze policy. |
| VM → renderer | Existing VM integration expanded by the WASM approach | install/sync collision calls `getDrawableCollisionState`; effects call `penLineRuns`. |
| renderer → VM compiled runtime | No direct import; API remains one-way | Renderer only exposes API/Event, although that API is an implicit correctness prerequisite for ABI6 features. |
| Three copies of opcode/feature validation across VM/packager/protector | Defensive duplication against drift, but also a maintenance cost | `src/compiler/opcode-semantics.js`, protector `opcode-semantics.mjs`, packager `opcode-semantics.js`; feature contracts are similarly distributed. |

### 4.2 Obvious Cross-Layer / Reverse Dependencies

1. **The VM compiled bridge depends on specialized renderer capabilities.** `compiled-project.js::syncTargetCollision()` assumes `renderer.getDrawableCollisionState()`, `CollisionGeometryChanged`, and `UseHighQualityRenderChanged`; `ABI5EffectsChannel` assumes `penLineRuns()` and fencing APIs. **[Fact]** This is not an import cycle, but the VM has a strong runtime contract with renderer extensions.
2. **Generic `RenderedTarget` knows about the compiled runtime.** Several setters directly check `runtime._compiledProjectTargetBinding`, increment `_compiledStateRevision`, and mark collision dirty. **[Fact]** This is cross-layer coupling in which a lower-level compilation approach leaks back into the generic sprite model.
3. **The physical WASM ABI layout is hard-coded in two repositories.** obfuscator `wasm-kernel.mjs` defines memory/exports; VM `wasm-thread-adapter.js` and `compiled-project.js` read exact names, numbers, and slot/command shapes. **[Fact]**
4. **Naming lag hides architectural evolution.** The ABI6 contract still lives in `wasm-kernel-abi5.js`, and adapter/effects/benchmark naming still says ABI5. **[Fact]** New contributors could mistakenly fix ABI6 bugs in what appears to be an “old path” or assume ABI5 compatibility is still expected.

### 4.3 High-Coupling Areas (Ranked by Risk, Not Defect)

| Area | Why it is complex | Typical change impact |
| --- | --- | --- |
| `ir-lowering.mjs` | 167 opcodes, type proofs, private allocator, suspension, host commands, and specializations are centralized here | A semantic change can affect memory, WASM stack shape, adapter behavior, and tests. |
| `wasm-thread-adapter.js` | Handles input, commands, host promises, scheduler interaction, reconciliation, monitors, and collision | New host/query/yield behavior crosses multiple statuses and fast paths. |
| `abi5-effects-channel.js` | Validates untrusted/encoded transport while also preserving order, batching Pen/motion/clone work, and providing fallback | Any effects ABI or renderer API change requires updates on both sides. |
| `compiled-project.js::install()` | Creates/removes targets, caches hats, handles monitor/cloud state, collision, and resource cleanup | Lifecycle omissions or clone-timing errors can appear as non-local bugs. |
| Feature/opcode contract mirrored three ways | Prevents incompatible artifacts but must be kept synchronized on every ABI evolution | Drift fails at packaging or loading time and is difficult to localize across repositories. |

## 5. Core Abstractions and Lifecycles

| Abstraction | Defined in | Creator | Consumer | Lifecycle / responsibility |
| --- | --- | --- | --- | --- |
| `IntermediateRepresentation` / `IntermediateScript` / `IntermediateStack(Block)` / `IntermediateInput` | `scratch-vm/src/compiler/intermediate.js` | `IRGenerator`, `IROptimizer` | JSGenerator, serializer, `compiledProject.create`, lowerer (serialized form) | Shared IR for a hat/procedure; after build-input generation only the portable tree is retained. |
| compiled IR v2 build input | `compiled-project.js::create` | VM | protector, packager validation | Contains `irProgram`, `privateState`, `publicProject`, contracts, extension requirements; cross-process/repository data boundary. |
| opcode semantics contract | VM/packager/protector validators | VM emits | packager/protector validate | Version+hash for the 167-opcode inventory; prevents the lowerer from silently accepting the wrong IR. |
| runtime feature contract | `wasm-kernel-abi5.js` / `abi6-features.*` | VM builds input; protector builds artifact | packager, loader, WASM shape validator | ABI6 + mandatory Wasm features + effects/collision contract; checked before loading. |
| protected artifact | `pipeline.mjs::protectCompiledProject` | protector | VM `save/load`, packager | `program/effects/publicProject/diagnostics`; later wrapped in zip + integrity manifest. |
| target ordinal + target handle | `compiled-project.js::install`, WASM ABI | install/WASM | adapter/effects channel | ordinal is original layout identity; handle is runtime identity including clones; clone state is deep-copied internally. |
| `WasmThreadAdapter` | `wasm-thread-adapter.js` | Blocks-cached `startingFunction` | Thread/Sequencer | One WASM thread handle per started hat Thread; adapts `runSlice` continuations into the existing generator-based scheduler. |
| `RunStatus` | `wasm-thread-adapter.js` and `wasm-kernel.mjs` | WASM `runSlice` | adapter | Private PC/suspension protocol; determines JS host call, yield, wait, and effects-drain behavior. |
| `ABI5EffectsChannel` | `abi5-effects-channel.js` | `compiledProject.install` | adapter, Runtime tick-end | Reads reusable views from fixed WASM effects memory, performs batching/fallback, and maintains statistics. Name is legacy; runtime is ABI6. |
| collision descriptor/state | `scratch-render/CollisionGeometry.js`, `RenderWebGL.getDrawableCollisionState` | Skin/renderer | VM bridge → WASM | Immutable conservative alpha geometry and position-independent transform; invalidated by skin/quality changes. |
| compiled target binding | Internal object in `compiled-project.js::install`, stored on `runtime._compiledProjectTargetBinding` | install | `Runtime`, `RenderedTarget`, Pen, adapter | Event subscription, flush/dirty/diagnostics/dispose; also a coupling point between generic runtime and compiled layer. |

## 6. Performance, Boundaries, Fallbacks, and Test Map

### 6.1 Key Hot Paths and Optimization Intent

| Hot path | Current strategy | WASM-specific? |
| --- | --- | --- |
| `runSlice` scheduling | Private PC and thread/call/argument arenas in WASM; `BUDGET` reduces host returns inside warp | Implementation is mainly WASM-specific; the scheduling-boundary idea is reusable |
| JS↔WASM input | `scriptInputMask`, 23-slot numeric snapshot, stage clock fast path, global input epoch/cache, reused `DataView/Float64Array` | Transport implementation is specific; on-demand/batched input is general |
| Command transport | Compact command readers, packed opcode reads, batched command flushing | ABI-export form is specific; batched event/command transport is general |
| Pen/motion | `effects.wasm` memory queue, ordered events, line runs, safe motion/deferred clone graphic motion | Wasm-to-Wasm channel is specific; renderer line-run batching is general |
| Renderer | Pen instancing, contiguous sprite instancing, reused uniform merge, draw-list append fast path | Does not depend on WASM; applicable to JS runtime |
| Collision | Alpha geometry → WASM packed BVH/tri-state; unknown falls back to renderer | Packed private BVH/tri-state bridge is WASM-oriented; geometry preprocessing is general |
| Strings/lists | Tagged scalar, private UTF-16, SIMD compare, fixed Unicode tables, owned buffers/list arena, numeric-list specialization | Strongly WASM-specific in the current implementation |
| Clone/thread churn | Target/thread handle reuse, incremental thread map, in-place disposed-target removal | Data-structure ideas can apply to JS runtime |

### 6.2 JS/WASM Boundaries and Fallbacks

| Boundary | Main path | Fallback / failure policy |
| --- | --- | --- |
| Feature support | ABI6 probes and runtime feature mask | Reject loading when mandatory features are unavailable; no JS fallback. |
| Opcode coverage | Lowerer fails closed; packager production gate | Incomplete coverage is only allowed through controlled benchmark/readable-WASM hooks; projects using unsupported opcodes are rejected. |
| Host/addon/compat | command/query + `HOST_COMMANDS/HOST_PENDING`, resumable PC | Execute JS primitive/extension; asynchronous result is written back before resume. |
| Pen/motion | Ordered effects + `penLineRuns` | When renderer/Pen/fencing/coordinate/drag conditions do not qualify, fall back to per-item JS mutation and synchronize actual results back to WASM. |
| Collision | WASM definite true/false | Unknown, render-only parameters, unregistered geometry, or unsafe geometry use renderer exact query. |
| External JS state | Revision-based reconciliation | When synchronization cannot be proven unnecessary, use full target snapshot/reconcile. |
| Online protection | `off`/`challenge` use the current path | `required` explicitly throws: in-WASM AEAD is not implemented; there is no insecure degradation. |

### 6.3 Test Coverage Map

| Category | Main location | Covered topics | Visible gaps / limitations |
| --- | --- | --- | --- |
| lowering/unit | `ScratchCN_v4_code-obfuscation/test/v2-protection.test.mjs` | WASM shape, input slots, PC/yield, host query, tail calls, numeric/string/list/clone behavior, private heap, Pen ordered effects, collision, trace | Very large and highly integrated test file; not equivalent to real browser/renderer combinations. |
| contract | obfuscator `opcode-conformance.test.mjs`; packager `abi6-features.test.js`, `compile-project.test.js` | Opcode partitioning, cross-target classes, VM/packager feature alignment, signing/extension-freeze/production gate | Maintenance cost of duplicated three-way contracts has not been eliminated by a single generated source. |
| VM integration | `scratch-vm/test/integration/tw_compiled_project.js`, `test/unit/compiler_abi5_effects_channel.js`, `compiler_wasm_effects_abi5.js` | Package loading, scheduler/clone/effects-channel shape | Filenames still say ABI5; static review did not confirm an independent regression test for every ABI6 fast path. |
| renderer correctness | `scratch-render/test/unit/{CollisionGeometryTests,PenSkinTests,RenderWebGLTests,SpriteInstancingTests}.js`, `integration/pen-line-runs-tests.js` | Conservative geometry proofs, readPixels equivalence between Pen runs and per-line rendering, fencing equivalence, instancing layout | Browser/GPU coverage is limited; pure unit tests cannot cover GPU-driver differences. |
| package/security | `tw-packager/test/packager/{sb3c,manifest-signing,csp,download-project,debug-trace}.test.js` | sb3c2, signature verification, CSP, loader, debug-trace CLI | Online `required` mode does not yet have a usable production implementation. |
| benchmark | `tw-packager/src/benchmark/abi5-runner.js` | PRNG, TerrainGen, Raycaster, Quicksort, List Search, String Parse, Pen TriFill, Pen Text; GPU fencing, AB/BA, profiling, collision smoke | Benchmark project is not in the repository; normal CI has no dedicated real-GPU worker, and documentation explicitly says this is not an automatic performance gate. |

### 6.4 How to Read the Archived Performance Evidence

**[Fact]** `tw-packager/docs/benchmark-results/abi6-2026-08-16/README.md` records three eight-metric AB/BA runs on Apple M5 / Chrome 151. The geometric mean over the six compute benchmarks is approximately 1.716–1.723; Quicksort is lowest at 0.853 and Pen Text Engine is lowest at 0.967. Pen Triangle Filler varies substantially across runs and is therefore excluded from the compute geometric mean.

**[Boundary]** These are experimental results for a fixed project, fixed machine, and fixed browser version. They cannot be generalized into an overall speedup for every Scratch project, browser, or GPU.

## 7. Git History by Architectural Stage, Not by Individual Commit

| Stage | Time / representative commits | Problem addressed and major changes | Continuing impact |
| --- | --- | --- | --- |
| A. Source-stripped JS protection prototype | 2026-08-05~08-08; VM `a5286f02`, packager `5dfdf70`, obfuscator `2f84697` | Established `.sb3c`/source-stripped packaging, integrity, JS obfuscator/accessor ABI, metadata cleanup | v1 executable-JS artifact was later removed; old docs/README still retain accessor/legacy narratives. |
| B. ABI2→ABI3: private WASM execution core | 2026-08-09; obfuscator `8131e7a`, `6a0627c`, `acbad38`, `3f36208`; corresponding VM commits | Added dynamic values, target isolation, clone foundations, suspension/yield state machine with private PC | The “WASM execution + JS host scheduler” split was established here; later synchronization complexity follows from it. |
| C. ABI4: capability requirements, strings, diagnosability | 2026-08-10~11; `ce62251`, VM `30c7c33b` | Fixed Unicode lowering, SIMD/bulk/tail/ref features, debug trace, readable sidecar, host-state crossing optimizations | ABI4 documentation still appears in `online-protection-v2.md` implementation status and no longer fully matches ABI6 code. |
| D. ABI5: effects module and hot Pen path | 2026-08-12~13; obfuscator `c0f1f53`, VM `4a7d981d`, renderer `f59879eb`, packager `890c644` | `effects.wasm`, fixed-memory command/effects transport, Pen line batching, complete opcode lowering, monitor/list/string/clone semantics | File and command names still say ABI5 after ABI6; effects bridge became a runtime core. |
| E. ABI5/6 host bridge: fewer crossings | 2026-08-13~14; obfuscator `1866fe8`, `04bb732`, `599a870`; VM `9033f6a6`, `6f8bfe35` | Batched input/target snapshots, compact exports, static cross-target internalization, query-suspension fixes | Adapter became both performance-critical and complex. |
| F. ABI6: collision and ordered effects | 2026-08-14~15; obfuscator `f9a2f8a`, `0166da3`, `d974562`, `988a06b`; VM `34a5f35c`; renderer `50d6f8ad` | Tri-state collision, packed BVH, renderer conservative geometry, ordered motion/clone graphic effects, position-only reconciliation, precise input masks | ABI 6 feature contract became fixed; renderer contract with compiled runtime deepened substantially. |
| G. Full-stack micro-optimization and qualification records | 2026-08-15~16; current HEADs near each repository, packager `fb0b7ea`/`35d7946` | Target/thread indexing, deferred/coalesced clone effects, uniform reuse/sprite instancing, benchmark grouping and qualification documentation | Performance optimizations are spread across four repositories; there is no unified performance design document or automatic real-GPU gate. |

## 8. Technical Debt, Workarounds, and Risk Hotspots

The following first describe current state. Suggestions are listed separately and should not be misread as already implemented fixes.

1. **ABI naming inconsistency.** **[Fact]** ABI constant is 6, while `wasm-kernel-abi5.js`, `abi5-effects-channel.js`, `benchmark:abi5`, and `instrumentABI5` remain in use. **[Suggestion]** First create a terminology/alias plan that does not change external compatibility, then perform one explicit migration rather than partially renaming things and increasing confusion.
2. **State revision is an intrusive workaround.** **[Fact]** Several generic `RenderedTarget` setters inspect `_compiledProjectTargetBinding`. **[Inference]** This was faster to implement than a generic “target mutation observer/state delta API,” but it makes the host model aware of a particular compilation approach. **[Suggestion]** If preparing upstream work, discuss a generic mutation-notification mechanism rather than upstreaming `_compiled*` private fields.
3. **Effects channel combines ABI validation, semantic recovery, GPU batching, and fallback.** **[Fact]** `ABI5EffectsChannel` is about 764 lines and has a multi-branch drain path. **[Inference]** This is a concentrated correctness/performance hotspot. **[Suggestion]** In the future, wire validation, effect interpretation, and renderer backend/fallback could be separated into smaller testable layers, but this should not be refactored casually in the current round.
4. **Hard-coded ABI layout / numeric opcodes.** **[Fact]** Lowerer/kernel/adapter contain command numbers, memory offsets, fixed capacities, and duplicated contracts across multiple repositories. **[Suggestion]** A shared generated schema/metadata layer could be considered later; simply moving constants without a versioned migration design risks creating a false “single source of truth.”
5. **Private/public dual authority is timing-sensitive.** **[Fact]** The clone-create comment explicitly notes that `targetWasCreated` fires before `RenderedTarget.makeClone()` finishes, so immediate reconciliation would overwrite the WASM clone state with temporary JS `(0,0)` values. **[Inference]** Similar risks can appear in drag, fencing, custom callbacks, and extension-driven mutation.
6. **Conservative invalidation of static specializations is complex.** **[Fact]** List/string ownership, numeric lists, Pen color, and cross-target/collision classes can degrade because of `compat`/addon/dynamic writes. **[Suggestion]** Every new optimization should include a semantic fallback test for proof invalidation rather than testing only the ideal fast path.
7. **Documentation state lags behind code.** **[Fact]** `tw-packager/docs/compiled-project-v2.md` still says “Packager 4” and “overall product is not a v2 security release”; the ABI4 status in `online-protection-v2.md` does not match current ABI6 code. These documents should be explicitly labeled historical/current to prevent them from being treated as release truth.
8. **Real performance gates are not automated.** **[Fact]** Benchmark docs explicitly state that the benchmark project is not in the repository and CI has no dedicated real-GPU worker. **[Suggestion]** Define traceable benchmark inputs and a hardware-runner contract before claiming a CI performance gate exists.

## 9. Common Modification Entry Points

| Common task | Read first | Check next | Pay special attention to |
| --- | --- | --- | --- |
| Add/change a Scratch opcode | `scratch-vm/src/compiler/enums.js`, `irgen.js`, `opcode-semantics.js` | `ir-lowering.mjs::Coverage` / corresponding lowering, three-way contract tests | Keep JS AOT semantics and WASM coverage/report aligned; unsupported cases must fail closed. |
| Change IR shape/fields | `intermediate.js`, `irgen.js`, `iroptimizer.js`, `compiled-ir-serializer.js` | `compiled-project.js::create`, `ir-lowering.mjs`, opcode validators | Portable serialization, debug trace, and whether JSGenerator still consumes the shape. |
| Change lowering / value representation | `ir-lowering.mjs` | `wasm-kernel.mjs`, `v2-protection.test.mjs` | Tagged values, allocator, clone/list ownership, Scratch conversion edge cases. |
| Change suspension/yield/warp | `ir-lowering.mjs` suspension/state-machine code | `wasm-thread-adapter.js::next`, `engine/sequencer.js`, `thread.js` | Scheduler-visible boundaries, broadcast-and-wait, promises, 500 ms warp timer. |
| Change clone behavior | lowerer `bindTarget`/clone lowering and `wasm-kernel.mjs::createBindTargetBody` | `compiled-project.js::targetWasCreated/targetWasRemoved`, `sprites/sprite.js`, `rendered-target.js` | Creation ordering, private deep copy, thread cleanup, handle reuse, drawable order. |
| Change variables/lists | `ir-lowering.mjs` slot/list headers, numeric/owned analysis | monitor/cloud sync in `compiled-project.js`, list cases in `v2-protection.test.mjs` | Visible monitors, cloud variables, clone sharing/ownership, string aliasing. |
| Change hostcall / compat / addon | lowerer `collectHostCalls`, HOST-command lowering | `wasm-thread-adapter.js::_runHostCall/_flushCommands`, `compat-block-utility.js` | Async result, query type, suspension/PC, external target mutation. |
| Change target-state synchronization | `compiled-project.js::reconcileTarget*` | `wasm-thread-adapter.js` staging/reconcile, `rendered-target.js` revisions | Authority direction, fencing/drag, renderer collision dirty state. |
| Change effects/Pen | `effects-kernel.wat`, `effects-kernel.mjs`, lowerer direct effects | `abi5-effects-channel.js`, `RenderWebGL.penLineRuns`, `PenSkin.drawLineRuns` | Exact import/export shape, event ordering, fallback, GPU readback tests. |
| Change collision/fallback | `CollisionGeometry.js`, `RenderWebGL.getDrawableCollisionState` | lowerer collision plan, `wasm-kernel.mjs` query/BVH, `compiled-project.js::syncTargetCollision` | Conservative proof, unknown→renderer fallback, quality/effect invalidation. |
| Change package/integrity | `tw-packager/compile-project.js`, `packager.js` | VM `virtual-machine.js`, `compiled-project-integrity.js`, `download-project.js` | Signing scope, sidecar exclusion, extension freeze, desktop-shell feature support. |
| Change benchmark/performance diagnostics | `src/benchmark/abi5-runner.js`, `scripts/benchmark-abi5.js`, `docs/abi5-benchmark.md` | `performance-monitor.js`, archived JSON schema/tests | Isolation of AOT vs WASM, AB/BA, GPU fencing, headless/SwiftShader bias. |

## 10. Classification: WASM-Specific, Potentially Upstreamable, Workarounds, and Possible Issues

### 10.1 WASM-Specific Designs / Optimizations

| Item | Assessment and evidence |
| --- | --- |
| Private linear memory, tagged scalar/string/list/PRNG, allocator | **[Fact]** Created by `ir-lowering.mjs` and `wasm-kernel.mjs`, with no memory export. This combines the protected execution model with the WASM implementation. |
| Linking `program.wasm` to import-free `effects.wasm` | **[Fact]** program imports only five functions from `sccn:abi5-effects`; effects has fixed 16-page memory. This module-to-module ABI directly uses Wasm linking. |
| SIMD string compare, bulk memory, tail calls, typed function references | **[Fact]** ABI6 feature contract/probes require these Wasm features; they are prerequisites for the current crossing/call/string optimizations. |
| Private PC / Wasm continuation | **[Fact]** Thread records and state-machine lowering keep resumable state inside WASM; JS only receives `RunStatus`. |
| Packed collision BVH and Wasm tri-state query | **[Fact]** Geometry is stored in private memory and queried by WASM as true/false/unknown; the data layout and execution point are WASM-specific. |

### 10.2 Implemented in WASM Today, but Potentially Applicable to TurboWarp’s JS Runtime

| Direction | Evidence and applicability |
| --- | --- |
| Precise per-script input dependencies + per-tick global cache | **[Fact]** `scriptInputMask` and global epoch/cache reduce staging; **[Inference]** JS AOT could also use IR read-sets to avoid unnecessary IO/clock snapshots. |
| Target revision / dirty reconciliation | **[Fact]** Currently exists to serve the JS/WASM bridge; **[Inference]** for pure JS AOT it could become a more general mechanism to avoid redundant target snapshots/renderer work, but only after defining a public mutation API. |
| Ordered Pen line-run batching, bulk fencing, instancing | **[Fact]** Renderer APIs themselves do not depend on WASM; `PenSkin` already has instanced/non-instanced fallback. **[Inference]** JS runtime Pen primitives could batch into the same API. |
| Contiguous-run sprite instancing, uniform reuse, draw-list append fast path | **[Fact]** Recent `RenderWebGL` optimizations do not depend on compiled artifacts. Suitable for independent renderer upstream discussion. |
| Precomputed collision geometry descriptors | **[Fact]** Descriptors are produced by the renderer; **[Inference]** even without the WASM BVH, JS runtime could use them for broad-phase/candidate pruning or caching. |
| Thread/target indexes, in-place removal, clone-ordering fast path | **[Fact]** Recent VM commits touch ordinary Runtime/target-array paths. These ideas can be evaluated independently of the WASM approach. |

### 10.3 Workarounds Created by the Current Architecture That Usually Should Not Be Upstreamed Directly

| Workaround | Why |
| --- | --- |
| `_compiledProjectTargetBinding`, `_compiledStateRevision`, `_projectCompiledPosition` spread through the generic VM model | Directly couples generic runtime code to one private artifact/runtime; upstream would be better served by general lifecycle/mutation contracts. |
| Manually synchronized exact export names/numbers/byte layouts across three repositories | Necessary for the protected artifact ABI, but inappropriate as an ordinary TurboWarp extension API. |
| Fixed effects capacity, numeric command IDs, special drain fences | Serve cross-Wasm/JS transport and particular hot workloads; upstream should start from a semantic batch API rather than exposing wire format. |
| Full feature requirement (tail calls, typed function references) as a player baseline | Reasonable for this protected executor but narrows general TurboWarp browser compatibility; should not be adopted upstream without product-level policy. |
| “Private authority → JS projection → fallback reverse reconciliation” | Synchronization compensation caused by dual execution domains; pure JS runtime does not have the same authority split. |

### 10.4 Findings That May Be Worth Separate TurboWarp Issues Later

The following are **[Suggestions]** for possible issues, not confirmed upstream bugs:

1. **Public/internal Pen bulk line-run API.** Goal: allow JS AOT or other producers to reuse `RenderWebGL.penLineRuns` / `PenSkin.drawLineRuns` while preserving per-line ordering and fallback semantics; readPixels equivalence tests should be the acceptance criterion.
2. **Contiguous sprite instancing with draw-order preservation.** `RenderWebGL._drawThese()` already implements contiguous-run instancing without a WASM dependency. Compatibility matrix, thresholds, and GPU performance data could be organized into a renderer performance issue/PR.
3. **Target mutation notification / revision API.** The current compiled private fields demonstrate a need: external mutations should be observable without embedding a private compiler mechanism into `RenderedTarget`. An issue should discuss generic event/version semantics only.
4. **Conservative collision descriptor API.** `CollisionGeometry` derives unknown/empty/outer/inner proofs from alpha masks. This could be discussed as an internal or experimental renderer API, provided its conservative contract and high-quality/effects invalidation are explicit.
5. **Reproducible GPU benchmark contract.** Formalize the one-pixel `readPixels` fence, fresh VM/renderer instances, AB/BA ordering, and headless limitations as benchmark methodology rather than as a claim that one engine is already faster.

## 11. Uncertainties and Things Static Review Cannot Confirm

1. **Real support matrix.** Feature probes provide recommended browser versions, but static reading cannot confirm the actual compatibility of every target browser, Safari/Firefox release channel, and desktop shell.
2. **Security strength.** The source confirms signing, source stripping, private memory, and fail-closed validation; source review alone cannot establish protection strength against a specific attacker. In particular, `required` online AEAD is not implemented and documentation also indicates that the production gate is not fully closed.
3. **Complete semantic equivalence.** Test coverage is broad, but the entire corpus, randomized differential testing, and every extension/browser/GPU combination were not executed. Full equivalence with TurboWarp cannot be claimed.
4. **Performance generalization.** Archived benchmarks cover only specific projects/devices. There is not enough evidence to claim the same benefit for ordinary projects, different clone densities, different GPUs, or low-end devices.
5. **Historical intent.** Git subjects and diffs show what changed, but cannot reliably reconstruct every design discussion, rejected alternative, or external-issue context.
6. **Purpose of the `TurboWarp-issue-18` snapshot.** Static directory structure shows copied code, but not whether it is an active working tree, submission material, or historical archive.

## 12. Conclusion-to-Source Index

| Conclusion | Type | Direct source |
| --- | --- | --- |
| Runtime ABI is 6, requires four Wasm feature groups, effects ABI is 4 | Fact | `scratch-vm/src/compiler/wasm-kernel-abi5.js`; `obfuscator/src/abi6-features.mjs`; `packager/src/packager/abi6-features.js` |
| VM and JS AOT share the same IR frontend | Fact | `scratch-vm/src/compiler/compile.js::generateIR/compile.toIR`; `intermediate.js` |
| Build input splits public/private state and uses executable-target order | Fact | `scratch-vm/src/compiler/compiled-project.js::create/collectPrivateState/createPublicProject` |
| Protected package is validated before modules are instantiated | Fact | `virtual-machine.js::loadCompiledProjectPackage`; `compiled-project.js::decodeProgram`; `compiled-project-integrity.js` |
| program imports only from the effects module, not host state | Fact | `obfuscator/src/wasm-kernel.mjs::EFFECT_IMPORTS/createProgramWasm`; `compiled-project.js::validateWasmShape` |
| Threads remain driven by the original Sequencer | Fact | `compiled-project.js::install`, `wasm-thread-adapter.js::next`, `engine/thread.js` |
| Ordered effect batching and fallback are a main performance/semantic boundary | Fact | `abi5-effects-channel.js::drain/_drainOrdered/_fallbackReason`; `RenderWebGL.js::penLineRuns` |
| Collision is renderer geometry + WASM tri-state + renderer fallback | Fact | `CollisionGeometry.js`; `RenderWebGL.js::getDrawableCollisionState`; `compiled-project.js::syncTargetCollision`; `wasm-kernel.mjs` |
| Recent optimization focuses on fewer crossings, effects batching, and collision/renderer work | Fact | Git history across the four repositories from 2026-08-12~16 (for example `1866fe8`, `4a7d981d`, `50d6f8ad`, `97c32b0f`, `29a6b9e`) |
| ABI5 naming is legacy rather than the actual runtime ABI | Fact | ABI constants are all 6 while files/commands still include `abi5`; see the feature contracts and filenames above |
| Dual authority is a major maintenance risk | Inference | The combination of reconcile/bind logic in `compiled-project.js`, `wasm-thread-adapter.js`, `abi5-effects-channel.js`, and revision/projection logic in `rendered-target.js` |
| Real GPU performance is not an existing CI gate | Fact | CI/qualification split in `tw-packager/docs/abi5-benchmark.md`; `abi6-2026-08-16/README.md` |

## 13. Chinese–English Terminology Reference

| Chinese | English | Recommended meaning in this project |
| --- | --- | --- |
| 编译项目 / 受保护项目 | compiled / protected project | `.sb3c2` produced from the original `.sb3`; not TurboWarp’s normal JS AOT output. |
| 构建输入 | build input | Portable IR v2 data package exported by the VM and consumed by the protector. |
| 降低 / 下放 | lowering | Mapping IR into WASM script bodies, memory layout, and host/effect ABIs. |
| 中间表示 | intermediate representation (IR) | `IntermediateRepresentation` and its portable serialized form. |
| 公共项目 | public project | Visible project data stored in `project.json` for resources/monitors/extensions. |
| 私有状态 | private state | Variables/lists/strings and other state stored in `program.wasm` private memory. |
| 状态所有权 | state authority / ownership | Which side is semantically authoritative for a state value at a given moment. |
| 调和 / 回填 | reconciliation | Staging external JS/renderer changes into WASM, or writing actual host results back into WASM after fallback. |
| 投影 | projection | Writing values already determined by WASM/effects into JS target/renderer without repeating host callbacks/fencing. |
| 效果通道 | effects channel | `program.wasm → effects.wasm → JS` transport for motion/Pen/graphics; source name still says ABI5. |
| 宿主调用 | host call / host command | Command or query that must be handled by JS VM/extension/renderer. |
| 暂停 / 续跑 | suspension / resume | WASM private PC stops at a host/wait/yield boundary and is later resumed by the adapter. |
| 运行切片 | run slice | One invocation of `runSlice(threadHandle, budget)`. |
| target ordinal | target ordinal | Stable original-target index aligned with IR/private layout. |
| target handle | target handle | WASM runtime identity; clones allocate/reuse separate handles. |
| 命令泵 | command pump | Protocol in which WASM produces host commands and the adapter reads/executes them in batches. |
| 精确输入掩码 | precise input mask | External input set actually required by a script, derived from IR/lowering. |
| 保守碰撞几何 | conservative collision geometry | Outer/inner rectangles derived from renderer alpha masks; unknown must fall back. |
| 三态查询 | tri-state query | Definite true, definite false, or unknown; unknown must not be treated as false. |
| 围栏 | fencing | Renderer semantics that constrain target/line coordinates to the visible stage area. |
| 快路径 / 回退 | fast path / fallback | Batch execution when proof conditions hold; otherwise use original JS semantics and reconcile the result. |
| 批处理 | batching | Combining commands, Pen lines, or sprite draws to reduce crossings, draw calls, or GC overhead. |
| 差分追踪 | differential trace | Comparing block/state events between an isolated interpreter and WASM using debug traces. |
| 上游 | upstream | The main TurboWarp project; issue candidates in this report are discussion directions only. |

---

### Appendix: Review Method

The static review covered package metadata, key source files, related tests/benchmarks, and Git history from 2026-08-05 through 2026-08-16 across the four active sub-repositories. No existing source code was modified, and no unexecuted browser test/benchmark was reported as passing.

This review additionally ran `npm test` in `ScratchCN_v4_code-obfuscation`: 125 of 129 tests passed, 4 were explicitly skipped because they cover removed ABI4 paths, and 0 failed (about 10.2 seconds). This validates only the protector’s Node test suite; it does not replace the full scratch-vm, renderer, packager, or real-GPU benchmark suites.


</details>
